h/views/api_auth.py

```diff
@@ -31,6 +32,43 @@ def __init__(self, context, request):
     @view_config(request_method='GET',
                  renderer='h:templates/oauth/authorize.html.jinja2')
     def get(self):
+        return self._authorize()
+
+    @view_config(request_method='GET',
+                 request_param='response_mode=web_message',
+                 renderer='h:templates/oauth/authorize.html.jinja2')
+    def get_web_message(self):
+        response = self._authorize()
+
+        if isinstance(response, HTTPFound):
```

Comment:
The `oauthlib` API is not designed to support the behaviour of `web_message`. It is designed to return a response object that should (theoretically) be directly rendered, which is why it returns headers, a body, and a status code.
They also don't have any support for hooking in different response modes. So the way that Nick and I ended up with was to use `oauthlib` as it is intended to be used, and then add additional functionality around it which handles the different behaviour when using `web_message`.
The only ugly part of this is that we need to parse the redirect URI to read out the authorization code, there is, unfortunately, no other way to get the authorization code. 

---

> To double-check my understanding: would I be correct in thinking that this depends on the fact that _authorize returns an HTTPFound in the case of a successful authorization, but raises one when the user isn’t logged in?

It is implemented in a way where it returns the successful response (`HTTPFound`), but raises an unexpected, error response.